go/src/lib.rs

```diff
@@ -1,9 +1,172 @@
+}
+
+/// The glide client.
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, RedisErrorFFI> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
+            }
+        })?;
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
+            }
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime.block_on(GlideClient::new(request)).map_err(|err| {
+        let msg = CString::new(err.to_string()).unwrap();
+        RedisErrorFFI {
+            message: msg.into_raw(),
+            error_type: ErrorType::ConnectionError,
+        }
+    })?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+/// Creates a new client to the given address. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+///
+/// The returned `ConnectionResponse` should be manually freed by calling `free_connection_response`, otherwise a memory leak will occur. It should be freed whether or not an error occurs.
+///
+/// # Safety
+///
+/// * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes.
+/// * `connection_request_len` must not be greater than `isize::MAX`. See the safety documentation of [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
+    connection_request_bytes: *const u8,
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            error: Box::into_raw(Box::new(err)) as *const RedisErrorFFI,
```

Comment:
Should you wrap this line with unsafe?
```suggestion
            error: unsafe { Box::into_raw(Box::new(err)) as *const RedisErrorFFI },
```
If yes, line 123 too

Reply:
In general, it's safe to create raw pointers in Rust. Dereferencing is unsafe, however.